pilot/pkg/serviceregistry/kube/controller/endpoints.go

```diff
@@ -181,24 +183,29 @@ func (e *endpointsController) onEvent(curr interface{}, event model.Event) error
 func (e *endpointsController) buildIstioEndpoints(endpoint interface{}, host host.Name) []*model.IstioEndpoint {
 	endpoints := make([]*model.IstioEndpoint, 0)
 	ep := endpoint.(*v1.Endpoints)
+	key := kube.KeyFunc(ep.Name, ep.Namespace)
 	for _, ss := range ep.Subsets {
 		for _, ea := range ss.Addresses {
 			pod := e.c.pods.getPodByIP(ea.IP)
 			if pod == nil {
-				// This means, the endpoint event has arrived before pod event. This might happen because
-				// PodCache is eventually consistent. We should try to get the pod from kube-api server.
 				if ea.TargetRef != nil && ea.TargetRef.Kind == "Pod" {
-					pod = e.c.pods.getPod(ea.TargetRef.Name, ea.TargetRef.Namespace)
+					podkey := kube.KeyFunc(ea.TargetRef.Name, ea.TargetRef.Namespace)
```

Comment:
Not sure about the utility of this part of double checking.  If getPodByIP returns nil using pod informer cache, checking it again right after likely will get the same result.

Reply:
I think it's useful, the first check is the pod cache which must make its way through the informer *and* the queue, so if it's out of order in the queue this may he useful

Reply:
At first I thought that too, but then I saw this:

https://github.com/istio/istio/blob/a72e83c4785a5d65f24e0baef2808ea49a20e0a3/pilot/pkg/serviceregistry/kube/controller/pod.go#L166

It appears to be doing a fallback check already.

Reply:
Won't the `if ! exist` part trigger?

I think it's very clear we need to add some extensive testing around this area so we don't have to guess

Reply:
Maybe there is a confusion here, but I was referring to that L196 just repeat what was in pod.go:166.

Reply:
It is little different to what pod cache's `getPodByIP` does. It just returns if it is not there in the pod cache - does not go to informer cache https://github.com/istio/istio/blob/master/pilot/pkg/serviceregistry/kube/controller/pod.go#L159.

This actually checks fall back informer existence check if it is not available in pod cache. 

We could fold this in to that method to avoid doing it here though

Reply:
Thanks the pointing that out, I was confused indeed.